File: flowcatalyst-typescript/packages/domain-core/src/unit-of-work.ts
/**
 * Unit of Work Pattern
 *
 * Ensures that entity state changes, domain events, and audit logs are
 * committed atomically within a single database transaction.
 *
 * **This is the ONLY way to create a successful Result.** The Result.success()
 * factory requires a token that only UnitOfWork has access to. This guarantees
 * that:
 * - Domain events are always emitted when state changes
 * - Audit logs are always created for operations
 * - Entity state and events are consistent (atomic commit)
 *
 * Usage in a use case:
 * ```typescript
 * async execute(cmd: CreateEventTypeCommand, ctx: ExecutionContext): Promise<Result<EventTypeCreated>> {
 *     // Validation - can return failure directly
 *     if (!isValid(cmd)) {
 *         return Result.failure(UseCaseError.validation('INVALID', 'Invalid input'));
 *     }
 *
 *     // Create aggregate
 *     const eventType = new EventType({ ... });
 *
 *     // Create domain event
 *     const event = new EventTypeCreated(ctx, { eventTypeId: eventType.id, ... });
 *
 *     // Atomic commit - only way to return success
 *     return this.unitOfWork.commit(eventType, event, cmd);
 * }
 * ```
 *
 * Note: This UnitOfWork is for control plane operations only. Event and
 * DispatchJob services are "leaf" operations that do not use this pattern
 * (to avoid circular dependencies).
 */

import type { DomainEvent } from './domain-event.js';
import type { Result } from './result.js';

/**
 * Aggregate entity interface.
 *
 * Entities must have an `id` field that is a TSID string.
 */
export interface Aggregate {
	readonly id: string;
}

/**
 * Unit of Work interface for atomic control plane operations.
 *
 * Implementations must:
 * 1. Execute all operations within a single database transaction
 * 2. Create the domain event record
 * 3. Create the audit log entry
 * 4. Only return Result.success() using the internal token
 */
export interface UnitOfWork {
	/**
	 * Commit an entity change with its domain event atomically.
	 *
	 * Within a single database transaction:
	 * 1. Persists or updates the aggregate entity
	 * 2. Creates the domain event in the events table
	 * 3. Creates the audit log entry
	 *
	 * If any step fails, the entire transaction is rolled back.
	 *
	 * @param aggregate - The entity to persist (must have id field)
	 * @param event - The domain event representing what happened
	 * @param command - The command that was executed (for audit log)
	 * @returns Success with the event, or Failure if transaction fails
	 */
	commit<T extends DomainEvent>(aggregate: Aggregate, event: T, command: unknown): Promise<Result<T>>;

	/**
	 * Commit a delete operation with its domain event atomically.
	 *
	 * Within a single database transaction:
	 * 1. Deletes the aggregate entity
	 * 2. Creates the domain event in the events table
	 * 3. Creates the audit log entry
	 *
	 * @param aggregate - The entity to delete (must have id field)
	 * @param event - The domain event representing the deletion
	 * @param command - The command that was executed (for audit log)
	 * @returns Success with the event, or Failure if transaction fails
	 */
	commitDelete<T extends DomainEvent>(aggregate: Aggregate, event: T, command: unknown): Promise<Result<T>>;

	/**
	 * Commit multiple entity changes with a domain event atomically.
	 *
	 * Use this for operations that create or update multiple aggregates,
	 * such as provisioning a service account (Principal + OAuthClient + Application).
	 *
	 * Within a single database transaction:
	 * 1. Persists or updates all aggregate entities
	 * 2. Creates the domain event in the events table
	 * 3. Creates the audit log entry
	 *
	 * If any step fails, the entire transaction is rolled back.
	 *
	 * @param aggregates - The entities to persist (each must have id field)
	 * @param event - The domain event representing what happened
	 * @param command - The command that was executed (for audit log)
	 * @returns Success with the event, or Failure if transaction fails
	 */
	commitAll<T extends DomainEvent>(aggregates: Aggregate[], event: T, command: unknown): Promise<Result<T>>;
}

/**
 * Abstract base class for UnitOfWork implementations.
 *
 * Provides the success token to implementations for creating successful results.
 */
export { RESULT_SUCCESS_TOKEN, type ResultSuccessToken } from './result.js';

File: flowcatalyst-go/internal/platform/common/unit_of_work.go
package common

import "context"

// UnitOfWork defines the interface for atomic operations that persist
// aggregates, emit domain events, and create audit logs transactionally.
//
// This is the ONLY way to return a successful Result from a use case.
// The Commit methods return Result using the unexported newSuccess constructor,
// which guarantees that:
//  1. Domain events are always emitted when state changes
//  2. Audit logs are always created for operations
//  3. Entity state and events are consistent (atomic commit)
//
// Use cases MUST call one of the Commit methods to return success.
// Direct calls to Result.Success() are not possible (it's unexported).
//
// Example usage in a use case:
//
//	func (uc *CreateEventTypeUseCase) Execute(
//	    ctx context.Context,
//	    cmd CreateEventTypeCommand,
//	    execCtx *common.ExecutionContext,
//	) common.Result[*EventTypeCreated] {
//	    // Validation - can return failure directly
//	    if !isValid(cmd) {
//	        return common.Failure[*EventTypeCreated](common.ValidationError(...))
//	    }
//
//	    // Create aggregate
//	    eventType := &eventtype.EventType{...}
//
//	    // Create domain event
//	    event := &EventTypeCreated{...}
//
//	    // Atomic commit - ONLY way to return success
//	    return uc.unitOfWork.Commit(ctx, eventType, event, cmd)
//	}
type UnitOfWork interface {
	// Commit persists an aggregate with its domain event atomically.
	//
	// Within a single database transaction:
	//  1. Persists or updates the aggregate entity (upsert by ID)
	//  2. Creates the domain event in the events collection
	//  3. Creates the audit log entry
	//
	// If any step fails, the entire transaction is rolled back.
	//
	// Parameters:
	//   - ctx: Context for the operation (includes timeout, cancellation)
	//   - aggregate: The entity to persist (must have an ID field)
	//   - event: The domain event representing what happened
	//   - command: The command that was executed (for audit logging)
	//
	// Returns:
	//   - Success with the event if commit succeeds
	//   - Failure with error if commit fails
	Commit(ctx context.Context, aggregate any, event DomainEvent, command any) Result[DomainEvent]

	// CommitDelete deletes an aggregate with its domain event atomically.
	//
	// Within a single database transaction:
	//  1. Deletes the aggregate entity
	//  2. Creates the domain event in the events collection
	//  3. Creates the audit log entry
	//
	// If any step fails, the entire transaction is rolled back.
	CommitDelete(ctx context.Context, aggregate any, event DomainEvent, command any) Result[DomainEvent]

	// CommitAll persists multiple aggregates with a domain event atomically.
	//
	// Use for operations that affect multiple aggregates, such as:
	//   - Provisioning (creates Application + ServiceAccount + OAuthClient)
	//   - Bulk operations
	//
	// Within a single database transaction:
	//  1. Persists or updates all aggregate entities
	//  2. Creates the domain event in the events collection
	//  3. Creates the audit log entry
	//
	// If any step fails, the entire transaction is rolled back.
	CommitAll(ctx context.Context, aggregates []any, event DomainEvent, command any) Result[DomainEvent]

	// CommitWithClientID is like Commit but also sets the clientId on the event.
	// Use for multi-tenant operations where events are scoped to a client.
	CommitWithClientID(ctx context.Context, aggregate any, event DomainEvent, command any, clientID string) Result[DomainEvent]
}

// AggregateRoot is an optional interface that aggregates can implement
// to provide collection name and ID extraction.
type AggregateRoot interface {
	// AggregateID returns the unique identifier for this aggregate.
	AggregateID() string

	// CollectionName returns the MongoDB collection name for this aggregate type.
	CollectionName() string
}

// Auditable is an optional interface that commands can implement
// to customize how they are serialized for audit logging.
type Auditable interface {
	// ToAuditJSON returns the JSON representation for audit logging.
	// Use this to redact sensitive fields like passwords.
	ToAuditJSON() string
}
